### PR TITLE
ecds: create shared base class for DynamicFilterConfigProviderImpl

### PR DESCRIPTION
**Pull request by tbarrella**
cc @kyessenov @lambdai

Commit Message:
ecds: create shared base class for DynamicFilterConfigProviderImpl

Signed-off-by: Taylor Barrella <tabarr@google.com>

Additional Description: Part of https://github.com/envoyproxy/envoy/issues/14696#issuecomment-897855672. The plan is to later change the signature of `onConfigUpdate` (as in #14717) and refactor more so that `DynamicExtensionConfigProviderBase` is no longer a template
Risk Level: Low
Testing: Existing
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
#14696

### CODE REVIEW

File: source/common/filter/config_discovery_impl.h [hunk 1]
```diff
@@ -65,21 +126,11 @@ class DynamicFilterConfigProviderImpl : public DynamicFilterConfigProvider {
     absl::optional<Envoy::Http::FilterFactoryCb> config_{};
   };
 
-  FilterConfigSubscriptionSharedPtr subscription_;
-  const absl::flat_hash_set<std::string> require_type_urls_;
   // Currently applied configuration to ensure that the main thread deletes the last reference to
   // it.
-  absl::optional<Envoy::Http::FilterFactoryCb> current_config_{absl::nullopt};
+  absl::optional<Envoy::Http::FilterFactoryCb> current_config_;
```
Comment (rojkov) [inline source/common/filter/config_discovery_impl.h:131]: nit: I'd probably keep it as it is to indicate `current_config_` isn't going to be inited in ctor.

File: source/common/filter/config_discovery_impl.cc [hunk 2]
```diff
@@ -28,84 +28,42 @@ void validateTypeUrlHelper(const std::string& type_url,
...
   validateTypeUrlHelper(type_url, require_type_urls_);
 }
 
-const std::string& DynamicFilterConfigProviderImpl::name() { return subscription_->name(); }
-
-absl::optional<Envoy::Http::FilterFactoryCb> DynamicFilterConfigProviderImpl::config() {
-  return tls_->config_;
-}
-
-void DynamicFilterConfigProviderImpl::onConfigUpdate(Envoy::Http::FilterFactoryCb config,
```
Comment (rojkov) [inline source/common/filter/config_discovery_impl.cc:71]: What's the reason for moving these implementations to the header? Is `DynamicFilterConfigProviderImpl` going to become a template?

Comment (tbarrella) [reply source/common/filter/config_discovery_impl.cc:71]: Sorry yes, that's the plan

**Review by rojkov — APPROVED** [review]
Thanks! LGTM

**Pull request merged**